### ISSUE TITLE
I can't comfortably browse the list of extensions using the keyboard and orca.

### ISSUE THREAD

**Issue by jvesouza** [microsoft/vscode#92483]
Issue Type: <b>Bug</b>

1. Press ctrl+shift+x.
2. type something like for example java.
Orca announces the number of extensions found as expected.
3. Try using the arrows to browse the list of extensions.
Note that the orca is silent.

If I am not mistaken previously it was possible to press the tab key and browse the list of found extensions.

@isidorn Am I doing something wrong?
Thanks.

VS Code version: Code - Insiders 1.44.0-insider (ef75c4bb445e5021065cf0012cfa2c66e55bfa16, 2020-03-11T09:19:54.444Z)
OS version: Linux x64 5.5.8-arch1-1

**Comment by isidorn** [comment]
@jvesouza this works just fine for me. Please note that before step 3. you have to press tab to move from the input box to the extensions list and then use the up / down arrow. If I do it like that Orca reads everything just fine.
If you think this behavior is counter intuitive we could look into changing it.

I have tried both on Linux and Mac and we properly set the aria-labels and the screen reader reads them.

**Comment by jvesouza** [comment]
@isidorn Really if I press the tab key I can navigate the list of extensions, but I got the impression that it is not exactly a list because when pressing the down arrow it seems that I hear information from the same extension.
In practice it looks like I need to press the down arrow two or three times to find the next extension in the list.

**Comment by isidorn** [comment]
@jvesouza for me that is not the case. Pressing down once nicely navigates and reads the next item. Are you sure that maybe you are not using some bad "java" example where there are duplicated extensions for some weird reason?

Can we maybe check with other Orca users if they reproduce this? @ajborka @webczat

**Comment by jvesouza** [comment]
I tried it with python, php and dotnet and the result was the same.
Let's see what other orca users say.

**Comment by ajborka** [comment]
@jvesouza , @isidorn , the extensions view works as expected. I tested with Orca master (3.37.1 pre), Ubuntu 20.04, and vscode that was updated with today's build. Keep in mind that you will have to turn on focus mode to navigate the less verbose version of the extensions list.

**Comment by jvesouza** [comment]
@ajborka @isidorn  HAHA! It works as expected if I turn on focus mode after press the tab key.
My question is: Why orca is switching to browse mode after I type some text and press the tab key?
Thanks.

**Comment by ajborka** [comment]
It might have something to do with aria-role or focus problems.

**Comment by jvesouza** [comment]
@isidorn @ajborka It would be great if we didn't have to switch to focus mode manually.

**Comment by isidorn** [comment]
@jvesouza agree. If you have some tips on what we should change let me know.
Currently the extensions list has the role of `list`.

**Comment by ajborka** [comment]
@isidorn , @jvesouza : Try the aria-role=application attribute. Here is an article that explains its use and why you would need to use it. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Application_Role

**Comment by jvesouza** [comment]
@isidorn Unfortunately, I have no idea what could be changed to remedy this situation as I am not very knowledgeable about how the orca works.
Maybe @joanmarie can help.

**Comment by joanmarie** [comment]
I think what @ajborka suggested (re using the ARIA `application` role) might solve the problem.

**Comment by isidorn** [comment]
@joanmarie thanks. But on what elemnt should the application role be set on? On the top element in the application right? Setting that role does not make sense to be done locally on some list elements as far as I understand.

**Comment by joanmarie** [comment]
In theory Orca should switch to focus mode if it's on the top element or a descendant widget.

In practice, we may discover that doesn't happen -- in which case I'll fix it in Orca. :)

**Comment by isidorn** [comment]
@joanmarie ok makes sense. I have pushed the change such that the vscode window now has the `application` role.
@jvesouza can you please try it out tomorrow, and we can see if additional changes are needed with Orca.

fyi @bpasero since I added the role in the `layout.ts`. I noticed that we add other classes there so that felt like the best place. Let me know if you would like me to move it.

**Comment by bpasero** [comment]
@isidorn maybe better to workbench itself: https://github.com/microsoft/vscode/blob/fd3e5b204bd837ce550cfcbfe264cf2eeb4f77ba/src/vs/workbench/browser/workbench.ts#L326

**Comment by isidorn** [comment]
@bpasero makes sense, pushed it.

**Comment by webczat** [comment]
I'd suggest to also make sure nvda works with this last change. I believe it also reacts to the application role.

**Comment by joanmarie** [comment]
@isidorn I just built VSCode master, see the `embedded` role (which is what ARIA's `application` role maps to in ATK/AT-SPI2). And when I do a search, press Tab, and then down arrow, Orca is presenting the extension as a whole. If I want to read the details (e.g. description and author), I need to switch to Orca's browse mode. So it looks like this is fixed.

Related aside: While I don't have a stable (or even preview) version of VSCode installed, here's my guess as to why Orca was in browse mode for @jvesouza: The list of extensions has the (ATK/AT-SPI2) role of `list` -- which suggests it's like an HTML `ul` or `ol` (i.e. not interactive). But the user interaction suggests this list is interactive. I don't know if it would make sense to use the ARIA `listbox` role or not.

Thanks (again!) for your work on VSCode accessibility for Orca users.

**Comment by isidorn** [comment]
@webczat makes sense, I just tried it with NVDA and seem to work fine. However I am not a screen reader user so for better coverage we will see tomorrow since we have quite some users using VS Code insiders and NVDA.

@joanmarie great, than optimistcally closing this issue, and if the issue reaparrs we can reopen. Thanks!

**Comment by jvesouza** [comment]
@isidorn  It is fixed, great job!
I still have a small question:
How to abandon the list of extensions without taking an action?
Let's assume the following scenario:
1. I press ctrl + shift + x. 
2. I type java and press the tab key. 
3. I use the arrows to browse the list of extensions and I cannot find an extension.

How to abandon this action? For me the ideal would be to press the escape key to return to the point I was in when I pressed ctrl + shift + x.

**Comment by webczat** [comment]
note that the extension view is like an explorer view or any other view. for me it would be illogical to make this view special.

**Comment by isidorn** [comment]
Great that it is fixed, adding verified label.

Yes we should not make the extensions view special, it is just like the explorer.
Well you can do what I do, Ctrl + B to hide the sidebar, that way the focus will move to the Editor.
Or you can Cmd + shift + E for example to move to a differnt siedbar view (in this example Explorer)

**Comment by ajborka** [comment]
@isidorn  this is fixed for me. Nice work! @jvesouza  you can always press ctrl+shift+e to return to the explorer. However, there are other paths to take, it just depends on your preference. Isn't there a command for open last used editor?

**Comment by isidorn** [comment]
@ajborka just closing the sidebar focused the last opened editor. And you can close the sidebar via `ctrl + B`

**Comment by jvesouza** [comment]
@isidorn @ajborka  Ctrl+b is a nice shortcut!

**Comment by jvesouza** [comment]
@isidorn It looks like the problem is back.

I can no longer browse the list of extensions using the 2020-03-18T05: 46: 17.067Z code-insiders but I can browse the list using the 2020-03-16T07:12:23.431Z code-insiders.

**Comment by isidorn** [comment]
@jvesouza the change I did was change the role of the extensons `list` to `listbox`. Because role `list` is not officially supported and was breaking inteli-sense. More details here https://github.com/microsoft/vscode/issues/92719#issuecomment-599569272

@joanmarie might it be that Orca does not handle the `listbox` role well here?

**Comment by joanmarie** [comment]
> @joanmarie might it be that Orca does not handle the `listbox` role well here?

Maybe? :slightly_smiling_face: Let me update my build and take a look.

**Comment by jvesouza** [comment]
> > @joanmarie might it be that Orca does not handle the `listbox` role well here?
> 
> Maybe? slightly_smiling_face Let me update my build and take a look.

Just out of curiosity, it would be no less difficult to download directly from https://code.visualstudio.com/docs/?dv=linux64&build=insiders?

**Comment by ajborka** [comment]
@isidorn , @joanmarie  confirmed on the break. It worked the day before yesterday (Tuesday). Both insiders and exploration are broke.

**Comment by joanmarie** [comment]
This is another case of Orca not getting events for the selection/focus changes. In other words, Orca is not being told the user is moving in the list. Because Orca has no idea the user is moving in the list, it has nothing to present. This is a variation of the Settings tree item problem and not something I can fix in Orca. We need to figure out why Chromium is not telling Orca that focus/selection is changing.

**Comment by ajborka** [comment]
@joanmarie , @isidorn , @jvesouza  the problem persists across multiple desktop environments. @joanmarie , this is a check based on other Chromium problems discussed on the Orca list. So, not sure if the lack of events in Gnome/Mate desktop environments are at fault, Chromium, or vscode. @isidorn , any ideas on where to go from here?

**Comment by joanmarie** [comment]
From a git bisect of VS Code, I can confirm that commit 08097a230 is the first bad commit. Before that change, Orca gets object:state-changed:focused events for each list item you arrow to. That's why Orca speaks the items after you move.

After that change, Orca gets absolutely no events from items in the list box. Which, again, is why Orca is silent.

@isidorn Are you setting the activedescendant as the user navigates in the listbox?

**Comment by isidorn** [comment]
@joanmarie yes I am setting the `activedescendent` as the user navigates in the listbox.

I think the actual issue is that `listbox` is the wrong role to use here. `listbox` is something like a native HTML selectbox, which should not be used for just a plan list of items. https://www.w3.org/WAI/PF/aria/roles#listbox
Children of `listbox` should be `option` and children of `list` are `listitems`. So for the VS Code use case we should use the role `list`. Howeveer that causes an NVDA issue. Thus I suggest that I simply revert to usign `tree`. I will also ask NVDA folk for clarification on the `list` issue https://github.com/microsoft/vscode/issues/92719

**Comment by isidorn** [comment]
Thinking about it more, we should use role `tree` and `treeitem` for all our trees. `list` and `listitem` for all our lists and `listbox` and `option` for our intelisense suggest box (you can trigger it in editor via ctrl+ space)

**Comment by webczat** [comment]
hmm, what is used for intellisense suggestion now, btw? it is not announced as a list or anything.

**Comment by isidorn** [comment]
@webczat `listbox` is used in the latest insiders that you are using. Does that experience make sense?

**Comment by webczat** [comment]
hmm, likely... as in... not too much difference except there is now a role announcement. but, I am wondering why orca announces it as something, that could be translated as... list button... or something similar. seems a bit out of place. :dagger:

**Comment by joanmarie** [comment]
Aside: Please prefer https://w3c.github.io/aria/ for your ARIA spec needs. The PF Working Group (under which ARIA used to live) ceased to exist years ago. And while some roles are pretty stable, others can get modified as we (ARIA Working Group) update the spec(s).

**Comment by isidorn** [comment]
@joanmarie Thanks. Bookmarking it.
I have pushed a fix for this.  So VS Code insiders from Monday should have the fix. 
Thanks all for reporting and providing feedback.
